macOSで`S`, `T`に色がついてしまう

![image](https://user-images.githubusercontent.com/10496148/231645352-823dd489-5e5e-4c5d-a0fd-c9f9ffbc84a8.png)


---

```
     LSCOLORS            The value of this variable describes what color to use for which attribute when colors are enabled with CLICOLOR or COLORTERM.  This string is a concatenation of pairs of the format fb, where f is the
                         foreground color and b is the background color.

                         The color designators are as follows:

                               a     black
                               b     red
                               c     green
                               d     brown
                               e     blue
                               f     magenta
                               g     cyan
                               h     light grey
                               A     bold black, usually shows up as dark grey
                               B     bold red
                               C     bold green
                               D     bold brown, usually shows up as yellow
                               E     bold blue
                               F     bold magenta
                               G     bold cyan
                               H     bold light grey; looks like bright white
                               x     default foreground or background

                         Note that the above are standard ANSI colors.  The actual display may differ depending on the color capabilities of the terminal in use.

                         The order of the attributes are as follows:

                               1.   directory
                               2.   symbolic link
                               3.   socket
                               4.   pipe
                               5.   executable
                               6.   block special
                               7.   character special
                               8.   executable with setuid bit set
                               9.   executable with setgid bit set
                               10.  directory writable to others, with sticky bit
                               11.  directory writable to others, without sticky bit

                         The default is "exfxcxdxbxegedabagacad", i.e., blue foreground and default background for regular directories, black foreground and red background for setuid executables, etc.
```

--- a/srcs/print_utils.c
+++ b/srcs/print_utils.c
@@ -30,6 +30,86 @@ int	print_filename_body(const char*	name, t_quote_type qt) {
 #endif
 }
 
+// ** カラーリングについて **
+
+// [macOS]
+
+//  LSCOLORS The value of this variable describes what color to use for which attribute when colors are enabled with CLICOLOR or COLORTERM.  This string is a concatenation of pairs of the format fb, where f is the
+//
+//           foreground color and b is the background color.
+//
+//           The color designators are as follows:
+//                 a     black
+//                 b     red
+//                 c     green
+//                 d     brown
+//                 e     blue
+//                 f     magenta
+//                 g     cyan
+//                 h     light grey
+//                 A     bold black, usually shows up as dark grey
+//                 B     bold red
+//                 C     bold green
+//                 D     bold brown, usually shows up as yellow
+//                 E     bold blue
+//                 F     bold magenta
+//                 G     bold cyan
+//                 H     bold light grey; looks like bright white
+//                 x     default foreground or background
+//
+//           Note that the above are standard ANSI colors.  The actual display may differ depending on the color capabilities of the terminal in use.
+//
+//           The order of the attributes are as follows:
+//
+//                 1.   directory
+//                 2.   symbolic link
+//                 3.   socket
+//                 4.   pipe
+//                 5.   executable
+//                 6.   block special
+//                 7.   character special
+//                 8.   executable with setuid bit set
+//                 9.   executable with setgid bit set
+//                 10.  directory writable to others, with sticky bit
+//                 11.  directory writable to others, without sticky bit
+//
+//           The default is "exfxcxdxbxegedabagacad", i.e., blue foreground and default background for regular directories, black foreground and red background for setuid executables, etc.
+
+// [Linux]
+
+// # Below are the color init strings for the basic file types.
+// # One can use codes for 256 or more colors supported by modern terminals.
+// # The default color codes use the capabilities of an 8 color terminal
+// # with some additional attributes as per the following codes:
+// # Attribute codes:
+// # 00=none 01=bold 04=underscore 05=blink 07=reverse 08=concealed
+// # Text color codes:
+// # 30=black 31=red 32=green 33=yellow 34=blue 35=magenta 36=cyan 37=white
+// # Background color codes:
+// # 40=black 41=red 42=green 43=yellow 44=blue 45=magenta 46=cyan 47=white
+// #NORMAL 00 # no color code at all
+// #FILE 00 # regular file: use no color at all
+// RESET 0 # reset to "normal" color
+// DIR 01;34 # directory
+// LINK 01;36 # symbolic link. (If you set this to 'target' instead of a
+//  # numerical value, the color is as for the file pointed to.)
+// MULTIHARDLINK 00 # regular file with more than one link
+// FIFO 40;33 # pipe
+// SOCK 01;35 # socket
+// DOOR 01;35 # door
+// BLK 40;33;01 # block device driver
+// CHR 40;33;01 # character device driver
+// ORPHAN 40;31;01 # symlink to nonexistent file, or non-stat'able file ...
+// MISSING 00 # ... and the files they point to
+// SETUID 37;41 # file that is setuid (u+s)
+// SETGID 30;43 # file that is setgid (g+s)
+// CAPABILITY 30;41 # file with capability
+// STICKY_OTHER_WRITABLE 30;42 # dir that is sticky and other-writable (+t,o+w)
+// OTHER_WRITABLE 34;42 # dir that is other-writable (o+w) and not sticky
+// STICKY 37;44 # dir with the sticky bit set (+t) and not other-writable
+// # This is for files with execute permission:
+// EXEC 01;32
+
 int	print_filename(const t_file_batch* batch, const t_file_item* item) {
 	const t_global_option* option = batch->opt;
 	static bool	colored = false;
@@ -70,14 +150,29 @@ int	print_filename(const t_file_batch* batch, const t_file_item* item) {
 			color = YO_COLOR_DIR;
 		}
 		colored = true;
+#ifdef __MACH__
+	// 8.   executable with setuid bit set
+	// 9.   executable with setgid bit set
+	} else if ((item->st.st_mode & S_IXUSR) && (item->st.st_mode & S_ISUID)) {
+		// 実行可能 かつ setuid-ed
+		color = YO_COLOR_UID;
+		colored = true;
+	} else if ((item->st.st_mode & S_IXGRP) && (item->st.st_mode & S_ISGID)) {
+		// 実行可能 かつ setgid-ed
+		color = YO_COLOR_GID;
+		colored = true;
+#else
+	// SETUID 37;41 # file that is setuid (u+s)
+	// SETGID 30;43 # file that is setgid (g+s)
 	} else if (item->st.st_mode & S_ISUID) {
-		// 実行可能な uid が制限されている
+		// setuid-ed
 		color = YO_COLOR_UID;
 		colored = true;
 	} else if (item->st.st_mode & S_ISGID) {
-		// 実行可能な gid が制限されている
+		// setgid-ed
 		color = YO_COLOR_GID;
 		colored = true;
+#endif
 	} else if (item->st.st_mode & S_IXUSR) {
 		// オーナーが実行可能
 		color = YO_COLOR_EXE;

--- a/test_result.mac.txt
+++ b/test_result.mac.txt
@@ -20,62 +20,29 @@
 [32m[ OK ][m tty: -l srcs/
 [32m[ OK ][m tty: -l srcs/*
 [32m[ OK ][m tty: -l ./includes/../srcs/
---- /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.648Qg9V6	2023-04-13 12:44:21.000000000 +0900
-+++ /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.fRcL9bS4	2023-04-13 12:44:21.000000000 +0900
-@@ -16,7 +16,7 @@
- -rw-r--r--  1 corvvs  staff  12493  4 13 12:40 ./includes/../srcs/print_long_format.c
- -rw-r--r--  1 corvvs  staff   4244  4  9 20:05 ./includes/../srcs/print_utils.c
- -rw-r--r--  1 corvvs  staff   6873  4  8 01:27 ./includes/../srcs/printf.c
---rw-r--r--  1 corvvs  staff    720  4 13 12:42 ./includes/../srcs/time.c
-+-rw-r--r--  1 corvvs  staff    720  4 13 12:42 ./includes/../srcs/time.c
- -rw-r--r--  1 corvvs  staff   1462  4  7 13:01 ./includes/../srcs/xattr.c
- 
- ./includes/../srcs/b:
-[31m[FAIL] tty: -l ./includes/../srcs/*[m
+[32m[ OK ][m tty: -l ./includes/../srcs/*
 [32m[ OK ][m tty: -l ./includes/../srcs/*
 [32m[ OK ][m tty: Makefile
 [32m[ OK ][m tty: Makefile src/
 [32m[ OK ][m tty: libft libft
---- /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.648Qg9V6	2023-04-13 12:44:21.000000000 +0900
-+++ /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.fRcL9bS4	2023-04-13 12:44:21.000000000 +0900
-@@ -103,7 +103,7 @@
- -rw-r--r--  1 corvvs  staff    528  4 13 12:44 ft_isascii.o
- -rw-r--r--  1 corvvs  staff   1020  3 26 16:37 ft_isdigit.c
- -rw-r--r--  1 corvvs  staff    544  4 13 12:44 ft_isdigit.o
---rw-r--r--  1 corvvs  staff    973  3 26 16:37 ft_isprint.c
-+-rw-r--r--  1 corvvs  staff    973  3 26 16:37 ft_isprint.c
- -rw-r--r--  1 corvvs  staff    528  4 13 12:44 ft_isprint.o
- -rw-r--r--  1 corvvs  staff   1402  3 26 16:37 ft_itoa.c
- -rw-r--r--  1 corvvs  staff    912  4 13 12:44 ft_itoa.o
-[31m[FAIL] tty: -l libft libft[m
+[32m[ OK ][m tty: -l libft libft
 [32m[ OK ][m tty: -R libft libft
---- /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.648Qg9V6	2023-04-13 12:44:21.000000000 +0900
-+++ /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.fRcL9bS4	2023-04-13 12:44:21.000000000 +0900
+--- /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.sy2lJW2Y	2023-04-13 13:05:52.000000000 +0900
++++ /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.4U0WIRaN	2023-04-13 13:05:52.000000000 +0900
 @@ -107,7 +107,7 @@
- -rw-r--r--  1 corvvs  staff    528  4 13 12:44 ft_isascii.o
+ -rw-r--r--  1 corvvs  staff    528  4 13 13:04 ft_isascii.o
  -rw-r--r--  1 corvvs  staff   1020  3 26 16:37 ft_isdigit.c
- -rw-r--r--  1 corvvs  staff    544  4 13 12:44 ft_isdigit.o
+ -rw-r--r--  1 corvvs  staff    544  4 13 13:04 ft_isdigit.o
 --rw-r--r--  1 corvvs  staff    973  3 26 16:37 ft_isprint.c
 +-rw-r--r--  1 corvvs  staff    973  3 26 16:37 ft_isprint.c
- -rw-r--r--  1 corvvs  staff    528  4 13 12:44 ft_isprint.o
+ -rw-r--r--  1 corvvs  staff    528  4 13 13:04 ft_isprint.o
  -rw-r--r--  1 corvvs  staff   1402  3 26 16:37 ft_itoa.c
- -rw-r--r--  1 corvvs  staff    912  4 13 12:44 ft_itoa.o
+ -rw-r--r--  1 corvvs  staff    912  4 13 13:04 ft_itoa.o
 [31m[FAIL] tty: -lR libft libft[m
 [32m[ OK ][m tty: -R
 [32m[ OK ][m tty: .
 [32m[ OK ][m tty: -R .
---- /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.648Qg9V6	2023-04-13 12:44:21.000000000 +0900
-+++ /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.fRcL9bS4	2023-04-13 12:44:21.000000000 +0900
-@@ -1,7 +1,7 @@
-                       '"'                   -                     NEW                   a.txt                 h''h                  [35mlink2[39;49m[0m                 [34mminimini[39;49m[0m              sl                    test_mac.sh           xc
- !                     '#                    --l                   OLD                   a\tb                  h'h'                  [35mlink3[39;49m[0m                 noones                [34msrcs[39;49m[0m                  test_mine_err.txt     {
- "'                    ''                    -l                    README.md             b.txt                 hello"world"          linux.txt             [34mobjs[39;49m[0m                  [30m[41msstk[39;49m[0m                  test_mine_out.txt     |
--#                     '\'                   :                     STK                   c.txt                 hello'world'          [35mls[39;49m[0m                    [34mpocket[39;49m[0m                t                     test_real_err.txt     }
-+#                     '\'                   :                     [30m[41mSTK[39;49m[0m                   c.txt                 hello'world'          [35mls[39;49m[0m                    [34mpocket[39;49m[0m                t                     test_real_err.txt     }
- $                     'h'h'                 ;                     [                     [34md[39;49m[0m                     [34mincludes[39;49m[0m              [31mlsls[39;49m[0m                  pocketdir             [34mtest1[39;49m[0m                 test_real_out.txt
- %                     'hello'               <                     \                     dir.c                 [34miunreadable[39;49m[0m           [35mm[39;49m[0m                     r                     [34mtest_field[39;49m[0m            test_result.linux.txt
- &                     (                     =                     ]                     [34mdocker[39;49m[0m                kkk                   m.c                   real                  [34mtest_field1[39;49m[0m           test_result.mac.txt
-[31m[FAIL] tty: -G[m
+[32m[ OK ][m tty: -G
 [32m[ OK ][m tty: -G srcs
 [32m[ OK ][m tty: -G includes
 [32m[ OK ][m tty: -G minimini
@@ -168,31 +135,9 @@
 [32m[ OK ][m file: -laRr test_field
 [32m[ OK ][m file: -laRt test_field
 [32m[ OK ][m file: -laRf test_field
---- /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.648Qg9V6	2023-04-13 12:44:29.000000000 +0900
-+++ /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.fRcL9bS4	2023-04-13 12:44:29.000000000 +0900
-@@ -34,7 +34,7 @@
- NEW
- OLD
- README.md
--STK
-+[30m[41mSTK[39;49m[0m
- [
- \
- ]
-[31m[FAIL] file: --color[m
+[32m[ OK ][m file: --color
 [32m[ OK ][m file: --color=auto
---- /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.648Qg9V6	2023-04-13 12:44:29.000000000 +0900
-+++ /var/folders/5q/_j05yjln22n75gbqwpnjvzz00000gn/T/tmp.fRcL9bS4	2023-04-13 12:44:29.000000000 +0900
-@@ -34,7 +34,7 @@
- NEW
- OLD
- README.md
--STK
-+[30m[41mSTK[39;49m[0m
- [
- \
- ]
-[31m[FAIL] file: --color=always[m
+[32m[ OK ][m file: --color=always
 [32m[ OK ][m file: --color=none
 [32m[ OK ][m tty: -l /Applications
 [32m[ OK ][m tty: -lG /Applications

